staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate.go

```diff
@@ -374,10 +397,19 @@ func (o AnnotateOptions) RunAnnotate() error {
 			return nil
 		}
 
-		return o.PrintObj(outputObj, o.Out)
+		return o.PrintObj(dataChangeMsg).PrintObj(outputObj, o.Out)
```

Comment:
`o.PrintObj.PrintObj` reads oddly to me. Since we can't actually cache the printer anymore, I'm wondering what the benefit is of storing a func on a field at all at this point. How about initializing the printer inline here, or perhaps using a regular method like so? (with a non-pointer receiver so that we're not mutating the original Operation on AnnotateOptions in the loop, which is another thing that bugs me currently)

```go
func (o AnnotateOptions) printObject(dataChangeMsg string, obj runtime.Object) error {
	o.PrintFlags.NamePrintFlags.Operation = dataChangeMsg
	printer, err := o.PrintFlags.ToPrinter()
	if err != nil {
		return err
	}
	return printer.PrintObj(obj, o.Out)
}
```

Reply:
Related to that, would it work to compare the final result to the original `info.Object` directly as runtime.Object at this point rather than tracking and setting the `dataChangeMsg` throughout this function?

Reply:
> `o.PrintObj.PrintObj` reads oddly to me. Since we can't actually cache the printer anymore, I'm wondering what the benefit is of storing a func on a field at all at this point. How about initializing the printer inline here, or perhaps using a regular method like so? (with a non-pointer receiver so that we're not mutating the original Operation on AnnotateOptions in the loop, which is another thing that bugs me currently)
> 
> ```go
> func (o AnnotateOptions) printObject(dataChangeMsg string, obj runtime.Object) error {
> 	o.PrintFlags.NamePrintFlags.Operation = dataChangeMsg
> 	printer, err := o.PrintFlags.ToPrinter()
> 	if err != nil {
> 		return err
> 	}
> 	return printer.PrintObj(obj, o.Out)
> }
> ```
> 
> 

But this ^ creates printobject per object and we might not be able to add yaml separator?

Reply:
Ah, I see. In the PR as it stands right now, we ARE reusing the printer, but we're relying on the fact that manipulating the `o.PrintFlags.NamePrintFlags.Operation` input state mutates the internal state of the already-initialized printer. I'm surprised that works (it's not what the existing label code does), and I don't think we can rely on it. 

Taking a step back, do we actually need to flag the unchanged state, or can we simply change the message to "annotation(s) modified" unconditionally? IIUC the only time they'd be unchanged is if you asked to apply annotations that already exist on the object with the exact same values. Until [very recently](https://github.com/kubernetes/kubernetes/commit/9feaf478c6a2d9982fa55086a856ccd2a414ecbe), that threw an error requiring `--overwrite` to be set. That PR didn't introduce any special no-op messaging.

Reply:
Updated to `ToPrinter`, to keep consistent with label.
PTAL thanks.